lib/Trace.js

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2015-2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import nodeUtils from './nodeUtils';
+
+class Node {
+    constructor(span, type, correlationId) {
+        this._span = span;
+        this._type = type;
+        this._correlationId = correlationId;
+        this._nodes = [];
+    }
+
+    getSpan() {
+        return this._span;
+    }
+
+    setType(type) {
+        this._type = type;
+    }
+
+    getNodes() {
+        return this._nodes;
+    }
+
+    getType() {
+        return this._type;
+    }
+
+    getCorrelationId() {
+        return this._correlationId;
+    }
+
+    setCorrelationId(correlationId) {
+        this._correlationId = correlationId;
+    }
+
+    toJSON() {
+        return nodeUtils.toJson(this);
+    }
+}
+
+class Trace {
+    constructor() {
+        this._nodes = [];
+    }
+
+    addNode(type, span, correlationId) {
+        let nodes = this._nodes;
+
+        const node = this.findNode(span.getParentId());
+        if (node) {
+            nodes = node.getNodes();
+        }
+
+        // TODO throw exception if span is already there?
+        nodes.push(new Node(span, type, correlationId));
+    }
+
+    setNodeType(type, span, correlationId) {
+        const node = this.findNode(span.getId());
+        if (node) {
+            node.setType(type);
+            node.setCorrelationId(correlationId);
+        } else {
+            console.log('Error setNode node not found');
+        }
+    }
+
+    /**
+     * Returns true if all spans associated with this trace are finished
```

Comment:
It doesn't return a boolean result - the comment and ideally the method name need to be changed to reflect its new behaviour.